### PR TITLE
feat : Upgrade keycloak 21.x 

### PR DESCRIPTION
**Pull request by BrayanMnz**
Close https://github.com/jhipster/generator-jhipster/issues/21819
Related with https://github.com/jhipster/generator-jhipster/pull/21938
---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by BrayanMnz** [comment]
May you trigger the checks? @DanielFran

**Comment by BrayanMnz** [comment]
As I said on the previous PR related to this 

The checks are failing due to exceptions like this

Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "http://localhost:9080/realms/jhipster/.well-known/openid-configuration": Unexpected end of file from server

I have seen that some people reported this before to JHipster but this is not something that I can say is caused by the Keycloak upgrade.

You can see here https://github.com/jhipster/generator-jhipster/issues/13733 where @mraible dealt with the same issue,
@mraible do you have any input here that you remember was helpful when you faced that issue?

**Comment by mraible** [comment]
@BrayanMnz It looks like you need to update snapshots with `npm run update-snapshots`.

**Comment by BrayanMnz** [comment]
done @mraible but, the issue that I've mentioned above still exists.

`Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "http://localhost:9080/realms/jhipster/.well-known/openid-configuration": Unexpected end of file from server`

File: generators/docker/templates/src/main/docker/keycloak.yml.ejs [hunk 1]
```diff
@@ -37,7 +37,7 @@ services:
       - 127.0.0.1:9080:9080
       - 127.0.0.1:9443:9443
     healthcheck:
-      test: ['CMD', 'curl', '-f', 'http://localhost:9080/realms/jhipster']
+      test: ['CMD',  'echo', '-n', 'GET', '/', 'HTTP/1.0\r\n\r\n', '|', 'nc', 'localhost/realms/jhipster', '9080', '$? eq 0']
```
Comment (mshima) [inline generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: Healthy check needs adjusts.
```suggestion
      test: ['CMD',  'echo', '-n', 'GET /realms/jhipster HTTP/1.1\r\n\r\n', '|', 'nc', 'localhost', '9080']
```

**Comment by mshima** [comment]
Failures looks related to keycloak not been ready.
Keycloak v20 is taking more than 50 to be healthy, v21 is takings - little more than 20 including download and build.

**Comment by BrayanMnz** [comment]
Hi @mshima can you enable the checks to run for the PR?

**Comment by BrayanMnz** [comment]
> Failures looks related to keycloak not been ready. Keycloak v20 is taking more than 50 to be healthy, v21 is taking little more than 20 including download and build.

you mean this failure?
`
Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "http://localhost:9080/realms/jhipster/.well-known/openid-configuration": Unexpected end of file from server`

**Comment by mshima** [comment]
> > Failures looks related to keycloak not been ready. Keycloak v20 is taking more than 50 to be healthy, v21 is taking little more than 20 including download and build.
> 
> you mean this failure? ` Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "http://localhost:9080/realms/jhipster/.well-known/openid-configuration": Unexpected end of file from server`

I think so.

**Comment by mraible** [comment]
@BrayanMnz It seems the health check you've switched to isn't working. I found a couple of resources that might help:

- [Enabling Keycloak Health checks](https://www.keycloak.org/server/health)
- [Keycloak V21 Docker Container Health Check Failing](https://stackoverflow.com/questions/75693830/keycloak-v21-docker-container-health-check-failing)

I also asked ChatGPT. It suggested `wget` or `nc`: https://chat.openai.com/share/a41f5514-8ca7-48cf-b348-b0a995498971

**Comment by BrayanMnz** [comment]
It is again failing with `Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "http://localhost:9080/realms/jhipster/.well-known/ope`

Do you suggest increasing the timeout for keycloak? @mshima

**Comment by mshima** [comment]
Still need to adjust the healthy check, it's not working yet.

**Comment by BrayanMnz** [comment]
I can see that it is working

Brayan Muñoz V.
________________________________
De: Marcelo Shima ***@***.***>
Enviado: Monday, June 5, 2023 9:48:45 PM
Para: jhipster/generator-jhipster ***@***.***>
Cc: Brayan Munoz Vargas ***@***.***>; Mention ***@***.***>
Asunto: Re: [jhipster/generator-jhipster] feat : Upgrade keycloak 21.x (PR #22429)


Still need to adjust the healthy check, it's not working yet.

**Comment by mshima** [comment]
It’s not working:
v20 log:
IMG_0029

v21 log ends without start the service:
IMG_0028

File: generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs [hunk 2]
```diff
@@ -2,6 +2,7 @@
   "id": "jhipster",
   "realm": "jhipster",
   "displayName": "JHipster",
+  "displayNameHtml" : "<div class=\"kc-logo-text\"><span>JHipster</span></div>",
```
Comment (mraible) [inline generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs:5]: I'm not sure this is necessary. What happens without it?

File: generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs [hunk 3]
```diff
@@ -1631,6 +1643,27 @@
...
+    "name" : "acr loa level",
+    "protocol" : "openid-connect",
+    "protocolMapper" : "oidc-acr-mapper",
+    "consentRequired" : false,
+    "config" : {
+    "id.token.claim" : "true",
+    "access.token.claim" : "true"
+    }
+    } ]
+    },
```
Comment (mraible) [inline generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs:1666]: The block looks like it needs formatting.

File: generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs [hunk 4]
```diff
@@ -2502,6 +2536,23 @@
...
+    "priority" : 70,
+    "config" : { }
+    }, {
+    "alias" : "webauthn-register-passwordless",
+    "name" : "Webauthn Register Passwordless",
+    "providerId" : "webauthn-register-passwordless",
+    "enabled" : true,
+    "defaultAction" : false,
+    "priority" : 80,
+    "config" : { }
```
Comment (mraible) [inline generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs:2555]: The block looks like it needs formatting.

File: generators/docker/templates/src/main/docker/config/realm-config/jhipster-realm.json.ejs [hunk 5]
```diff
@@ -1629,6 +1641,27 @@
...
+    "name" : "acr loa level",
+    "protocol" : "openid-connect",
+    "protocolMapper" : "oidc-acr-mapper",
+    "consentRequired" : false,
+    "config" : {
+    "id.token.claim" : "true",
+    "access.token.claim" : "true"
+    }
+    } ]
+    },
```
Comment (mraible) [inline generators/docker/templates/src/main/docker/config/realm-config/jhipster-realm.json.ejs:1664]: The block looks like it needs formatting.

File: generators/docker/templates/src/main/docker/config/realm-config/jhipster-realm.json.ejs [hunk 6]
```diff
@@ -2500,6 +2534,23 @@
...
+    "priority" : 70,
+    "config" : { }
+    }, {
+    "alias" : "webauthn-register-passwordless",
+    "name" : "Webauthn Register Passwordless",
+    "providerId" : "webauthn-register-passwordless",
+    "enabled" : true,
+    "defaultAction" : false,
+    "priority" : 80,
+    "config" : { }
```
Comment (mraible) [inline generators/docker/templates/src/main/docker/config/realm-config/jhipster-realm.json.ejs:2553]: Formatting.

File: generators/docker/templates/src/main/docker/keycloak.yml.ejs [hunk 7]
```diff
@@ -37,7 +37,7 @@ services:
       - 127.0.0.1:9080:9080
       - 127.0.0.1:9443:9443
     healthcheck:
-      test: ['CMD', 'curl', '-f', 'http://localhost:9080/realms/jhipster']
+      test: ['CMD',  'echo', '-n', 'GET /realms/jhipster HTTP/1.1\r\n\r\n', '|', 'nc', 'localhost', '9080']
```
Comment (mraible) [inline generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]:
```suggestion
      test: ['CMD', 'nc', '-z', '-w2', 'localhost', '9080']
```

Comment (mraible) [reply generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: In this case, we use the nc command with the -z option to scan for open ports, -w2 to specify a 2-second timeout, and localhost and 8080 as the host and port to check. If the connection fails or times out, nc will return a non-zero exit code, indicating an unhealthy state.

File: generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs (hunk 2 above)
Comment (BrayanMnz) [reply generators/docker-compose/templates/realm-config/jhipster-realm.json.ejs:5]: displayNameHtml, is just the setting for how the name will be displayed.

File: generators/docker/templates/src/main/docker/keycloak.yml.ejs (hunk 7 above)
Comment (BrayanMnz) [reply generators/docker/templates/src/main/docker/keycloak.yml.ejs]: Applied, please enable the checks to run again.

File: generators/docker/templates/src/main/docker/keycloak.yml.ejs [hunk 8]
```diff
@@ -37,7 +37,7 @@ services:
       - 127.0.0.1:9080:9080
       - 127.0.0.1:9443:9443
     healthcheck:
-      test: ['CMD', 'curl', '-f', 'http://localhost:9080/realms/jhipster']
+      test: ['CMD', 'nc', '-z', '-w2', 'localhost', '9080']
```
Comment (mraible) [inline generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: I believe we need to add `KC_HEALTH_ENABLED=true` to the environment and use the following for the test:

```
test: "bash /opt/keycloak/health-check.sh"
```

From https://twitter.com/ubaid4j/status/1666489377537822721

Comment (mraible) [reply generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: The `health-check.sh` script also needs to be added.

https://github.com/ubaid4j/api-gateway-with-multiple-oidc-clients/blob/main/src/main/docker/keycloak/health-check.sh

Comment (mraible) [reply generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: @BrayanMnz I tried to update this PR and push the changes to your branch, but I get a permission denied error. Maybe you can temporarily add me to your fork?

Comment (BrayanMnz) [reply generators/docker/templates/src/main/docker/keycloak.yml.ejs:40]: done!

File: generators/docker/files.mjs [hunk 9]
```diff
@@ -142,7 +142,16 @@ export const dockerFiles = {
     {
       condition: generator => generator.authenticationTypeOauth2,
       path: DOCKER_DIR,
-      templates: ['keycloak.yml', { file: 'config/realm-config/jhipster-realm.json', renameTo: () => 'realm-config/jhipster-realm.json' }],
+      templates: ['keycloak.yml',
+        {
+          file: 'config/keycloak-health-check.sh',
+          renameTo: () => 'realm-config/keycloak-health-check.sh'
```
Comment (mshima) [inline generators/docker/files.mjs:148]:
```suggestion
          renameTo: () => 'realm-config/keycloak-health-check.sh'
          transform: false,
```

**Comment by mraible** [comment]
The health check still seems to be failing. If I try to run it locally, it complains the file is not executable. After I `chmod +x`, it complains that the `timeout` command is not found. I'm not sure if this provides any insight.

**Comment by mshima** [comment]
@mraible great, it’s working.
Just need to replicate healthcheck to the compose file:
https://github.com/BrayanMnz/generator-jhipster/blob/b6fba84e80ae2bdfc8e5b89863e1ad55518e8415/generators/docker-compose/templates/docker-compose.yml.ejs#L127

**Review by mraible — APPROVED** [review]

**Comment by mraible** [comment]
Thanks for your contribution, @BrayanMnz!

**Comment by BrayanMnz** [comment]
Thank you to the whole team! @mraible 
Looking forward for the next one, let me know if you have something special that need an extra hand. 

(It's a holiday in my country today, sorry for being inactive)

**Comment by BrayanMnz** [comment]
bounty claimed https://opencollective.com/generator-jhipster/expenses/143406.

**Comment by DanielFran** [comment]
@BrayanMnz approved

**Pull request merged**